common/src/main/java/com/wynntils/core/chat/transcoder/CodedStyle.java

```diff
@@ -0,0 +1,238 @@
+            if (!this.isCompatibleWith(partBefore.getCodedStyle())) {
+                styleString.append(STYLE_PREFIX).append(ChatFormatting.RESET.getChar());
+            }
+        }
+
+        // 1. Color
+        if (color != CustomColor.NONE) {
+            Optional<ChatFormatting> chatFormatting = Arrays.stream(ChatFormatting.values())
+                    .filter(ChatFormatting::isColor)
+                    .filter(c -> c.getColor() == color.asInt())
+                    .findFirst();
+
+            if (chatFormatting.isPresent()) {
+                styleString.append(STYLE_PREFIX).append(chatFormatting.get().getChar());
+            } else {
+                styleString.append(STYLE_PREFIX).append(color.toHexString());
+            }
+        }
+
+        // 2. Formatting
+        if (bold) {
+            styleString.append(STYLE_PREFIX).append(ChatFormatting.BOLD.getChar());
+        }
+        if (italic) {
+            styleString.append(STYLE_PREFIX).append(ChatFormatting.ITALIC.getChar());
+        }
+        if (underlined) {
+            styleString.append(STYLE_PREFIX).append(ChatFormatting.UNDERLINE.getChar());
+        }
+        if (strikethrough) {
+            styleString.append(STYLE_PREFIX).append(ChatFormatting.STRIKETHROUGH.getChar());
+        }
+        if (obfuscated) {
+            styleString.append(STYLE_PREFIX).append(ChatFormatting.OBFUSCATED.getChar());
+        }
+
+        // 3. Click event
+        if (clickEvent != null) {
+            styleString
+                    .append(STYLE_PREFIX)
+                    .append("[")
+                    .append(owner.getParent().addClickEvent(clickEvent))
+                    .append("]");
+        }
+
+        // 4. Hover event
+        if (hoverEvent != null) {
+            styleString
+                    .append(STYLE_PREFIX)
+                    .append("<")
+                    .append(owner.getParent().addHoverEvent(hoverEvent))
+                    .append(">");
+        }
+
+        return styleString.toString();
+    }
+
+    public Style getStyle() {
+        if (styleCache != null) {
+            return styleCache;
+        }
+
+        Style reconstructedStyle = Style.EMPTY
+                .withBold(bold)
+                .withItalic(italic)
+                .withUnderlined(underlined)
+                .withStrikethrough(strikethrough)
+                .withObfuscated(obfuscated)
+                .withClickEvent(clickEvent)
+                .withHoverEvent(hoverEvent);
+
+        if (color != CustomColor.NONE) {
+            reconstructedStyle = reconstructedStyle.withColor(color.asInt());
+        }
+
+        styleCache = reconstructedStyle;
+
+        return reconstructedStyle;
+    }
+
+    private boolean isCompatibleWith(CodedStyle other) {
+        // A style is compatible with another style if:
+        // 1. If the color is non-null in this style, it is the same as the other style or other style's color is null.
```

Comment:
I'm not sure I get this logic. If we have `"red".withStyle(RED).append("blue".withStyle(BLUE))`, then you would say they are not "compatible", and thus you would issue a `§r`. But that will not be necessary..? In fact, I think you *only* need to issue a reset (basically, what you call "incompatible") if a style goes from true to false. Color can always be replaced.

(This comment ties in with how this is used in asString as well)

Reply:
I just fixed this before you types this out :). But yes, you are right.

Reply:
And no, there is a case where you have whatever color, then none color. Then you need to reset.